clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java

```diff
@@ -402,26 +403,38 @@ public void refreshCommittedOffsetsIfNeeded() {
         }
     }
 
-    @Override
-    public void close() {
+    public void close(long timeoutMs, long requestTimeoutMs) {
         // we do not need to re-enable wakeups since we are closing already
         client.disableWakeups();
+        long startTimeMs = time.milliseconds();
         try {
-            maybeAutoCommitOffsetsSync();
+            timeoutMs = Math.min(timeoutMs, requestTimeoutMs);
+            maybeAutoCommitOffsetsSync(timeoutMs);
+        } finally {
+            super.close();
+        }
 
-            Node coordinator;
-            long endTimeMs = time.milliseconds() + CLOSE_TIMEOUT_MS;
-            while ((coordinator = coordinator()) != null && client.pendingRequestCount(coordinator) > 0) {
-                long remainingTimeMs = endTimeMs - time.milliseconds();
-                if (remainingTimeMs > 0)
+        Node coordinator = coordinator();
+        long now = time.milliseconds();
+        long endTimeMs = now + Math.min(timeoutMs - (now - startTimeMs), requestTimeoutMs);
+        while (pendingAsyncCommits.get() > 0 || (coordinator != null && client.pendingRequestCount(coordinator) > 0)) {
```

Comment:
Seems the only thing we really care about is offset commits when shutting down. As long as we send the LeaveGroup, it's probably fine not to await its response (because of one of your previous patches). So because we now have the check for `pendingAsyncCommits`, I'm wondering if it's actually necessary to await all pending requests from the coordinator? At least if we keep the check, maybe we could ensure that we are not in the middle of a rebalance since that would unnecessarily delay shutdown.